service/storage_service.cc

```diff
@@ -1181,116 +1171,169 @@ class topology_coordinator {
-    // get admitted before global_tablet_token_metadata_barrier() is finished in
-    // handle_tablet_write_both_read_old().
-    future<> handle_tablet_streaming(group0_guard guard) {
-        std::vector<future<>> calls;
-        auto tm = get_token_metadata_ptr();
-        for (auto&& [table, tmap] : tm->tablets().all_tables()) {
-            co_await coroutine::maybe_yield();
-            for (auto&& [tablet, trinfo] : tmap.transitions()) {
-                co_await coroutine::maybe_yield();
-                auto gid = locator::global_tablet_id{table, tablet};
-                auto dst = trinfo.pending_replica.host;
-                slogger.info("Initiating tablet streaming of {} to {}", gid, trinfo.pending_replica);
-                calls.emplace_back(futurize_invoke([&] {
-                    return ser::storage_service_rpc_verbs::send_tablet_stream_data(&_messaging,
-                        netw::msg_addr(id2ip(dst)), _as, gid);
-                }).handle_exception([gid, dst](auto ep) {
-                    slogger.error("Failed to stream tablet {} to {}: {}", gid, dst, ep);
-                    std::rethrow_exception(ep);
-                }));
+            switch (trinfo.stage) {
+                case locator::tablet_transition_stage::allow_write_both_read_old:
+                    transition_to_with_barrier(locator::tablet_transition_stage::write_both_read_old);
+                    break;
+                case locator::tablet_transition_stage::write_both_read_old:
+                    transition_to_with_barrier(locator::tablet_transition_stage::streaming);
+                    break;
+                // The state "streaming" is needed to ensure that stale stream_tablet() RPC doesn't
+                // get admitted before global_tablet_token_metadata_barrier() is finished for earlier
+                // stage in case of coordinator failover.
+                case locator::tablet_transition_stage::streaming:
+                    if (advance_in_background(gid, tablet_state.streaming, "streaming", [&] {
+                        slogger.info("raft topology: Initiating tablet streaming of {} to {}", gid, trinfo.pending_replica);
+                        auto dst = trinfo.pending_replica.host;
+                        return ser::storage_service_rpc_verbs::send_tablet_stream_data(&_messaging,
+                                   netw::msg_addr(id2ip(dst)), _as, gid);
+                    })) {
+                        transition_to(locator::tablet_transition_stage::write_both_read_new);
+                    }
+                    break;
+                case locator::tablet_transition_stage::write_both_read_new:
+                    transition_to_with_barrier(locator::tablet_transition_stage::use_new);
+                    break;
+                case locator::tablet_transition_stage::use_new:
+                    transition_to_with_barrier(locator::tablet_transition_stage::cleanup);
```

Comment:
We'd schedule a cleanup operation here, no? Maybe leave a comment?

Reply:
Yes. I'll leave a fixme.